File: packages/python/examples/pytest/swag_labs_test.py
from os import getenv
from time import sleep

import pytest
from pytest import fixture, mark
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.support.expected_conditions import presence_of_element_located
from selenium.webdriver.support.wait import WebDriverWait

from alumnium import Provider

driver_type = getenv("ALUMNIUM_DRIVER", "selenium")


@fixture(autouse=True)
def login(al, driver, execute_script, navigate):
    al.learn("add laptop to cart", ["click button 'Add to cart' next to 'laptop' product"])
    al.learn("go to shopping cart", ["click link after 'Swag Labs' with a number text in it"])

    if driver_type == "appium-ios":
        al.learn(
            "sort products by lowest shipping cost",
            [
                "click generic element after 'Products' text",
                'click "Shipping (low to high)"',
            ],
        )
    else:
        al.learn(
            "sort products by lowest shipping cost",
            [
                "click combobox with options",
                'click option "Shipping (low to high)"',
            ],
        )

    navigate("https://www.saucedemo.com/")
    al.do("type 'standard_user' into username field")
    al.do("type 'secret_sauce' into password field")
    al.do("click login button")
    if driver_type == "appium-ios":
        wait = WebDriverWait(driver, 2)
        try:
            button = wait.until(
                presence_of_element_located(
                    (
                        "xpath",
                        "//XCUIElementTypeButton[@name='Not Now']",
                    )
                )
            )
            button.click()
        except TimeoutException:
            pass

    sleep(1)  # https://github.com/alumnium-hq/alumnium/issues/215
    yield
    execute_script("window.localStorage.clear()")

    al.clear_learn_examples()


def test_sorting(al):
    if al.model.provider == Provider.OLLAMA:
        pytest.xfail("Too hard for Mistral")

    products = {
        "Sauce Labs Backpack": 29.99,
        "Sauce Labs Bike Light": 9.99,
        "Sauce Labs Bolt T-Shirt": 15.99,
        "Sauce Labs Fleece Jacket": 49.99,
        "Sauce Labs Onesie": 7.99,
        "Test.allTheThings() T-Shirt (Red)": 15.99,
    }
    titles = list(products.keys())
    prices = list(products.values())

    # Default order is A-Z
    assert al.get("titles of products") == sorted(titles)

    al.do("sort products in descending alphabetical order")
    assert al.get("titles of products") == sorted(titles, reverse=True)

    al.do("sort products in ascending alphabetical order")
    assert al.get("titles of products") == sorted(titles)

    al.do("sort products by lowest price")
    assert al.get("prices of products (without money sign)") == sorted(prices)

    al.do("sort products by highest price")
    assert al.get("prices of products (without money sign)") == sorted(prices, reverse=True)


@mark.xfail(
    driver_type == "appium-ios",
    reason="https://github.com/alumnium-hq/alumnium/issues/132",
)
def test_checkout(al):
    model_provider = al.model.provider
    if model_provider == Provider.OLLAMA:
        pytest.xfail("Too hard for Mistral")
    if model_provider == Provider.MISTRALAI:
        pytest.xfail("Cannot figure out how to open cart")

    al.do("add onesie to cart")
    al.do("add backpack to cart")
    al.do("go to shopping cart")
    assert al.get("titles of products in cart") == [
        "Sauce Labs Onesie",
        "Sauce Labs Backpack",
    ]

    al.do("go to checkout")
    al.do("fill in first name - Al, last name - Um, ZIP - 95122")
    al.do("continue checkout")

    assert al.get("item total without tax (without money sign)") == 37.98
    assert al.get("tax amount (without money sign)") == 3.04
    assert al.get("total amount with tax (without money sign)") == round(37.98 + 3.04, 2)
    assert al.get("shipping information value") == "Free Pony Express Delivery!"

    al.do("finish checkout")

    al.check("thank you for the order message is shown")
    if model_provider != Provider.DEEPSEEK:
        al.check("big green checkmark is shown", vision=True)

File: packages/typescript/tests/system/swagLabs.test.ts
import { describe } from "vitest";
import { baseIt } from "./helpers.ts";

describe("Swag Labs", () => {
  const it = baseIt.override("setup", async ({ setup, skip }) => {
    return async (options) => {
      const result = await setup(options);
      const { al, driverId, model, $ } = result;

      await Promise.all([
        al.learn("add laptop to cart", [
          "click button 'Add to cart' next to 'laptop' product",
        ]),
        al.learn("go to shopping cart", [
          "click link after 'Swag Labs' with a number text in it",
        ]),
        al.learn("sort products by lowest shipping cost", [
          driverId === "appium-ios"
            ? "click generic element after 'Products' text"
            : "click combobox with options",
          driverId === "appium-ios"
            ? 'click "Shipping (low to high)"'
            : 'click option "Shipping (low to high)"',
        ]),
      ]);

      await $.navigate("https://www.saucedemo.com/");
      await al.do("type 'standard_user' into username field");
      await al.do("type 'secret_sauce' into password field");
      await al.do("click login button");
      await new Promise((resolve) => setTimeout(resolve, 1000));

      if (model.provider === "ollama") skip("Too hard for Mistral");

      return result;
    };
  });

  it("sorts products", async ({ expect, setup }) => {
    const { al } = await setup();
    const titles = [
      "Sauce Labs Backpack",
      "Sauce Labs Bike Light",
      "Sauce Labs Bolt T-Shirt",
      "Sauce Labs Fleece Jacket",
      "Sauce Labs Onesie",
      "Test.allTheThings() T-Shirt (Red)",
    ];
    const prices = [29.99, 9.99, 15.99, 49.99, 7.99, 15.99];

    expect(await al.get("titles of products")).toEqual(titles.toSorted());
    await al.do("sort products in descending alphabetical order");
    expect(await al.get("titles of products")).toEqual(
      titles.toSorted().toReversed(),
    );
    await al.do("sort products in ascending alphabetical order");
    expect(await al.get("titles of products")).toEqual(titles.toSorted());
    await al.do("sort products by lowest price");
    expect(await al.get("prices of products (without money sign)")).toEqual(
      prices.toSorted((a, b) => a - b),
    );
    await al.do("sort products by highest price");
    expect(await al.get("prices of products (without money sign)")).toEqual(
      prices.toSorted((a, b) => b - a),
    );
  });

  it("checks out", async ({ expect, setup, skip }) => {
    const { al, driverId, model } = await setup();

    if (driverId === "appium-ios")
      skip("https://github.com/alumnium-hq/alumnium/issues/132");
    if (model.provider === "mistralai")
      skip("Cannot figure out how to open cart");

    await al.do("add onesie to cart");
    await al.do("add backpack to cart");
    await al.do("go to shopping cart");
    expect(await al.get("titles of products in cart")).toEqual([
      "Sauce Labs Onesie",
      "Sauce Labs Backpack",
    ]);

    await al.do("go to checkout");
    await al.do("fill in first name - Al, last name - Um, ZIP - 95122");
    await al.do("continue checkout");
    expect(await al.get("item total without tax (without money sign)")).toBe(
      37.98,
    );
    expect(await al.get("tax amount (without money sign)")).toBe(3.04);
    expect(await al.get("total amount with tax (without money sign)")).toBe(
      41.02,
    );
    expect(await al.get("shipping information value")).toBe(
      "Free Pony Express Delivery!",
    );

    await al.do("finish checkout");
    await al.check("thank you for the order message is shown", {
      assert: expect.assert,
    });
    if (model.provider !== "deepseek")
      await al.check("big green checkmark is shown", {
        vision: true,
        assert: expect.assert,
      });
  });
});
